main/server.go

```diff
@@ -0,0 +1,313 @@
+	w.WriteHeader(statusCode)
+	return json.NewEncoder(w).Encode(v)
+}
+
+// ListTodos handles GET requests on /todos.
+// It requires no parameters and returns a list of todos.
+//
+// Example:
+//
+//   req: GET /todos/
+//   res: 200 [
+//          {"id": 1, "title": "Learn Go", "completed": false},
+//          {"id": 2, "title": "Buy bread", "completed": true}
+//        ]
+func ListTodos(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	res, err := todo.All(c)
+	if err != nil {
+		log.Errorf(c, "ListTodos: %v", err)
+		return err
+	}
+	return writeJson(w, http.StatusOK, res)
+}
+
+// NewTodo handles POST requests on /todos.
+// The request body must contain a JSON object with a Title field.
+// The status code of the response is used to indicate any error.
+//
+// Examples:
+//
+//   req: POST /todos/ {"title": ""}
+//   res: 400 empty title
+//
+//   req: POST /todos/ {"title": "Buy bread"}
+//   res: 201
+func NewTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	req := struct{ Title string }{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return badRequest{err}
+	}
+	t, err := todo.NewTodo(req.Title)
+	if err != nil {
+		return badRequest{err}
+	}
+	log.Infof(c, "Saving new todo: %v", t)
+	if err = t.Save(c); err != nil {
+		return err
+	}
+	return writeJson(w, http.StatusCreated, t)
+}
+
+// parseID obtains the id variable from the given request url,
+// parses the obtained text and returns the result.
+func parseID(r *http.Request) (int64, error) {
+	txt, ok := mux.Vars(r)["id"]
+	if !ok {
+		return 0, fmt.Errorf("todo id not found")
+	}
+	return strconv.ParseInt(txt, 10, 0)
+}
+
+// GetTodo handles GET requsts to /todos/{todoID}.
+// There's no parameters and it returns a JSON encoded todo.
+//
+// Examples:
+//
+//   req: GET /todos/1
+//   res: 200 {"id": 1, "title": "Buy bread", "completed": true}
+//
+//   req: GET /todos/42
+//   res: 404 todo not found
+func GetTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	id, err := parseID(r)
+	if err != nil {
+		return badRequest{err}
+	}
+	t, ok := todo.Get(appengine.NewContext(r), id)
+	if !ok {
+		return notFound{}
+	}
+	return writeJson(w, http.StatusOK, t)
+}
+
+// UpdateTodo handles PUT requests to /todos/{todoID}.
+// The request body must contain a JSON encoded todo.
+// The id property is optional, but if it is included, it must match
+// with the request path.
+//
+// Example:
+//
+//   req: PUT /todos/1 {"id": 1, "title": "Learn Go", "completed": true}
+//   res: 200
+//
+//   req: PUT /todos/2 {"id": 1, "title": "Learn Go", "completed": true}
+//   res: 400 inconsistent todo IDs
+func UpdateTodo(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	id, err := parseID(r)
+	if err != nil {
+		return badRequest{err}
+	}
+	var t todo.Todo
+	if err := json.NewDecoder(r.Body).Decode(&t); err != nil {
+		return badRequest{err}
+	}
+	if t.ID == 0 {
```

Comment:
Aaaaah, some of this weirdness is actually because i cribbed from you :)

I've gone ahead and removed anything about the `id` property in the payload, as it makes the API cleaner and conforms better with the todo-api spec.